File: memcache/memcache.cc
class Memcache {
public:
    Memcache() {
        // initialize your data structure here.
        items.clear();
    }
    
    int get(int curtTime, int key) {
        // Write your code here
        if (items.find(key) != items.end()) {
            if (items[key].second == -1 || items[key].second> curtTime) {
                return items[key].first;
            } else {
                items.erase(key);
            }
        }
        return INF;
    }
    
    void set(int curtTime, int key, int value, int ttl) {
        // Write your code here
        if (ttl == 0) {
            items[key] = make_pair(value, -1);
        } else {
            items[key] = make_pair(value, curtTime + ttl);
        }
    }
    
    void del(int curtTime, int key) {
        // Write your code here
        if (items.find(key) != items.end()) {
            items.erase(key);
        }
    }
    
    int incr(int curtTime, int key, int delta) {
        // Write your code here
        if (items.find(key) != items.end()) {
            if (items[key].second == -1 || items[key].second> curtTime) {
                return items[key].first += delta;
            } else {
                items.erase(key);
            }
        }
        return INF;
    }
    
    int decr(int curtTime, int key, int delta) {
        // Write your code here
        if (items.find(key) != items.end()) {
            if (items[key].second == -1 || items[key].second> curtTime) {
                return items[key].first -= delta;
            } else {
                items.erase(key);
            }
        }
        return INF;
    }
    
private:
    map<int, pair<int, int>> items;
    const int INF = 2147483647;
};

File: memcache/memcache.py
class Memcache:
    
    def __init__(self):
        # initialize your data structure here.
        self.items = dict()
        self.INF = 2147483647
    
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @return an integer
    def get(self, curtTime, key):
        # Write your code here
        if key in self.items:
            if self.items[key][1] == -1 or self.items[key][1] > curtTime:
                return self.items[key][0]
            else:
                self.items.pop(key)
        return self.INF
    
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @param {int} value an integer
    # @param {int} ttl an integer
    # @return nothing
    def set(self, curtTime, key, value, ttl):
        # Write your code here
        if ttl == 0:
            expire = -1
        else:
            expire = curtTime + ttl
        self.items[key] = [value, expire]
    
    # @param {int} curtTime an integer
    # @param {int} key an integer
    # @return nothing
    def delete(self, curtTime, key):
        # Write your code here
        if key in self.items:
            self.items.pop(key)

# @param {int} curtTime an integer
# @param {int} key an integer
# @param {int} delta an integer
# @return an integer
def incr(self, curtTime, key, delta):
    # Write your code here
    if key in self.items:
        if self.items[key][1] == -1 or self.items[key][1] > curtTime:
            self.items[key][0] += delta
                return self.items[key][0]
            else:
                self.items.pop(key)
    return self.INF

# @param {int} curtTime an integer
# @param {int} key an integer
# @param {int} delta an integer
# @return an integer
def decr(self, curtTime, key, delta):
    # Write your code here
    if key in self.items:
        if self.items[key][1] == -1 or self.items[key][1] > curtTime:
            self.items[key][0] -= delta
                return self.items[key][0]
            else:
                self.items.pop(key)
    return self.INF
